File: CharAnalysis_2_0_MATLAB/charLowess.m
function ys = charLowess(y, span, method)
% charLowess    Base-MATLAB replacement for smooth() (Curve Fitting Toolbox).
%   ys = charLowess(y, span)
%   ys = charLowess(y, span, method)
%
%   Replicates the behaviour of the Curve Fitting Toolbox function
%       smooth(y, span, method)
%   using only base MATLAB, so CharAnalysis can run without that toolbox.
%
%   INPUTS
%     y      : data vector (column or row; returned in the same orientation)
%     span   : smoothing window.
%              >= 1  interpreted as number of data points
%              <  1  interpreted as fraction of data length
%     method : 'lowess'  (default) locally-weighted linear regression
%              'rlowess' robust lowess (3 re-weighting iterations)
%              'moving'  simple moving average
%
%   OUTPUT
%     ys : smoothed vector, same size as y
%
%   charLowess first checks whether the Curve Fitting Toolbox is available.
%   If it is, smooth() is called directly, guaranteeing results identical
%   to CharAnalysis v1.1. If the toolbox is not available, a pure-MATLAB
%   implementation is used as a fallback.
%
%   ALGORITHM (pure-MATLAB fallback, lowess / rlowess)
%     For each point i, a neighbourhood of k points is selected.  At
%     interior points the window is centred on i.  At boundary points the
%     window is SHIFTED (not shrunk) to maintain exactly k points, matching
%     the behaviour of MATLAB's smooth() from the Curve Fitting Toolbox.
%     Each neighbour j receives a tricubic distance weight:
%
%         w_j = ( 1 - (|i-j| / d_max)^3 )^3
%
%     where d_max is the distance to the FURTHEST point in the window
%     (not the half-window radius), so weights adapt naturally to the
%     shifted boundary windows.
%
%     A weighted least-squares line is fitted and evaluated at i.
%     For 'rlowess', residuals are used to form bisquare robustness
%     weights, down-weighting outliers (3 passes).
%
%   EDGE HANDLING
%     MATLAB smooth() shifts the window at boundaries so every point is
%     fitted using exactly k neighbours.  The previous version of
%     charLowess shrunk the window instead, producing slightly different
%     values at record edges.  This version uses shifted windows to match
%     MATLAB smooth() behaviour and eliminate the edge discrepancy that
%     caused 2 peaks near the oldest part of the record to be missed.
%
%   CharAnalysis v2.0  -  Phase 1 modernisation

%% ── Input handling ───────────────────────────────────────────────────────
if nargin < 3 || isempty(method)
    method = 'lowess';
end
isRow = isrow(y);
y     = y(:);           % work on column vector throughout
n     = numel(y);

%% ── Convert span to integer window width ─────────────────────────────────
if span < 1
    k = max(3, round(span * n));    % fraction -> point count
else
    k = max(3, round(span));        % already a point count
end
k  = min(k, n);                     % cannot exceed data length
hw = floor(k / 2);                  % nominal half-window radius

%% ── Moving average ───────────────────────────────────────────────────────
if strcmpi(method, 'moving')
    ys = movmean(y, k, 'EndPoints', 'shrink');
    if isRow, ys = ys'; end
    return
end

%% ── Use Curve Fitting Toolbox smooth() if available ──────────────────────
% smooth() produces results identical to v1.1. The pure-MATLAB fallback
% below is used only on installations without the Curve Fitting Toolbox.
if license('test', 'Curve_Fitting_Toolbox')
    switch lower(method)
        case 'lowess'
            ys = smooth(y, k, 'lowess');
        case 'rlowess'
            ys = smooth(y, k, 'rlowess');
        case 'moving'
            ys = smooth(y, k, 'moving');
        otherwise
            ys = smooth(y, k, 'lowess');
    end
    if isRow, ys = ys'; end
    return
end

%% ── Pure-MATLAB fallback (no Curve Fitting Toolbox) ──────────────────────
if strcmpi(method, 'rlowess')
    nIter = 5;
else
    nIter = 1;
end

ys = y;
rw = ones(n, 1);        % robustness weights (all 1 for plain lowess)

for iter = 1:nIter
    for i = 1:n

        % ── Window selection: shift at boundaries to keep exactly k pts ──
        % This matches MATLAB smooth() edge behaviour.
        i0 = i - hw;
        i1 = i0 + k - 1;
        if i0 < 1
            i0 = 1;
            i1 = k;
        end
        if i1 > n
            i1 = n;
            i0 = n - k + 1;
        end

        xi = (i0:i1)';
        yi = y(i0:i1);
        ri = rw(i0:i1);

        % Tricubic weights: distance from i to each window point,
        % normalised by distance to the furthest point in the window.
        d    = abs(xi - i);
        dmax = max(d) + eps;
        tri  = (1 - (d ./ dmax) .^ 3) .^ 3;
        w    = tri .* ri;

        % Weighted least-squares linear fit
        sw   = sum(w);
        swx  = sum(w .* xi);
        swy  = sum(w .* yi);
        swx2 = sum(w .* xi .^ 2);
        swxy = sum(w .* xi .* yi);
        det  = sw * swx2 - swx ^ 2;

        if abs(det) < eps * max(abs([sw*swx2, swx^2]))
            ys(i) = swy / (sw + eps);   % near-singular: fall back to mean
        else
            a     = (swy * swx2 - swx * swxy) / det;
            b     = (sw  * swxy - swx * swy)  / det;
            ys(i) = a + b * i;
        end
    end

    % Bisquare robustness weights for rlowess
    if strcmpi(method, 'rlowess') && iter < nIter
        res = y - ys;
        s   = median(abs(res));
        if s < eps, break; end
        u  = res ./ (6 * s);
        rw = max(0, 1 - u .^ 2) .^ 2 .* (abs(u) < 1);
    end
end

if isRow, ys = ys'; end
end

File: CharAnalysis_2_0_R/R/charLowess.R
#' Locally-weighted linear regression matching MATLAB charLowess.m / smooth()
#'
#' Pure-R implementation of the shifted-window lowess algorithm from MATLAB's
#' \code{charLowess.m} v2.0, which itself replicates MATLAB's
#' \code{smooth(y, k, 'lowess')} from the Curve Fitting Toolbox.  Replaces
#' the previous wrapper around \code{\link[stats]{lowess}}, which used a
#' shrinking window at record boundaries and different tricubic weight
#' normalisation, causing systematic differences of ~1e-3 in \code{charBkg}.
#'
#' @param y     Numeric vector of values to smooth (NaN-free; bridge NaNs
#'   before calling and restore afterward, matching the MATLAB workflow).
#' @param x     Numeric vector of x-positions (same length as \code{y}).
#'   If \code{NULL}, \code{seq_along(y)} is used (index coordinates).
#' @param span  Smoothing span.
#'   \itemize{
#'     \item \code{span >= 1}: number of data points.
#'     \item \code{span < 1}: fraction of data length.
#'   }
#'   Converted to integer \code{k = max(3, round(span * n))} or
#'   \code{max(3, round(span))}, matching \code{charLowess.m} lines 63-68.
#' @param iter  Number of bisquare robustness passes after the initial fit.
#'   \itemize{
#'     \item \code{iter = 0}: plain lowess (1 total WLS pass, no robustness).
#'     \item \code{iter = 4}: robust lowess, matching MATLAB's \code{'rlowess'}
#'       which uses \code{nIter = 5} total iterations (1 initial + 4 updates).
#'   }
#'
#' @return Numeric vector of smoothed values, same length as \code{y}.
#'
#' @details
#'   ## Algorithm
#'   For each point \eqn{i}, a window of exactly \eqn{k} points is selected:
#'   the window is centred on \eqn{i} at interior points and SHIFTED (not
#'   shrunk) at record boundaries, maintaining \eqn{k} neighbours throughout.
#'   Each neighbour \eqn{j} receives a tricubic weight:
#'   \deqn{w_j = \left(1 - \left(\frac{|i-j|}{d_{\max}}\right)^3\right)^3}
#'   where \eqn{d_{\max}} is the distance to the \strong{furthest} point in
#'   the window (not the half-window radius).  A weighted least-squares line
#'   is fitted and evaluated at \eqn{i}.  This matches
#'   \code{charLowess.m} lines 109-147 exactly.
#'
#'   For \code{iter > 0} (\code{'rlowess'}), bisquare robustness weights are
#'   computed from the residuals after each WLS pass and multiplied into the
#'   tricubic weights for the next pass (matching \code{charLowess.m} lines
#'   151-157).
#'
#'   ## Why not stats::lowess()?
#'   \code{stats::lowess()} trims (shrinks) the window at boundaries and
#'   normalises tricubic weights by the half-window radius \eqn{hw}.  The two
#'   differences compound to ~1e-3 absolute error in \code{charBkg} for the CO
#'   dataset (500 yr / 15 yr = 33-point window, 17-point boundary zone).
#'
#' @seealso [char_smooth()], [char_thresh_local()]
char_lowess <- function(y, x = NULL, span = 0.1, iter = 0L) {

  n <- length(y)
  if (is.null(x)) x <- seq_len(n)

  # ---- NaN handling ----------------------------------------------------------
  # MATLAB's smooth() excludes NaN positions from local WLS fits and from the
  # bisquare residual median, then returns NA at the NaN centre positions.
  # The previous approach bridged NaN in CharSmooth before calling this
  # function, but that changes the bisquare weight trajectory on subsequent
  # passes, causing charBkg to diverge for records with gaps when using
  # method 2 (robust lowess).  Handling NaN here directly matches smooth().
  nan_mask <- is.na(y)
  has_nan  <- any(nan_mask)

  # ---- Convert span to integer window width ---------------------------------
  # Mirrors charLowess.m lines 63-68:
  #   if span < 1:  k = max(3, round(span * n))   % fraction -> point count
  #   else:         k = max(3, round(span))        % already a point count
  if (span < 1) {
    k <- max(3L, round(span * n))
  } else {
    k <- max(3L, round(span))
  }
  k  <- min(k, n)
  hw <- k %/% 2L                        # floor(k/2): nominal half-window

  n_iter <- 1L + as.integer(iter)       # total passes: 1 plain, 5 robust

  ys <- y
  # NaN positions permanently get weight 0 so they never contribute to a fit.
  rw <- ifelse(nan_mask, 0.0, 1.0)

  for (it in seq_len(n_iter)) {

    for (i in seq_len(n)) {

      # ---- NaN centre: pass through as NA ------------------------------------
      if (nan_mask[i]) { ys[i] <- NA_real_; next }

      # ---- Shifted window: always k points (charLowess.m lines 110-120) ----
      i0 <- i - hw
      i1 <- i0 + k - 1L
      if (i0 < 1L) { i0 <- 1L; i1 <- k        }
      if (i1 > n)  { i1 <- n;  i0 <- n - k + 1L }

      xi <- x[i0:i1]
      yi <- y[i0:i1]
      ri <- rw[i0:i1]

      # ---- Tricubic weights (charLowess.m lines 128-130) --------------------
      # Normalised by distance to the FURTHEST VALID point in the window.
      # NaN positions are excluded from dmax so they don't artificially
      # compress the weights of valid neighbours, matching smooth()'s
      # behaviour of treating NaN points as absent from the window.
      nan_in_win <- is.na(yi)
      d    <- abs(xi - x[i])
      d_valid <- d[!nan_in_win]
      if (length(d_valid) == 0L) { ys[i] <- NA_real_; next }
      dmax <- max(d_valid) + .Machine$double.eps
      tri  <- (1.0 - (d / dmax)^3)^3
      w    <- tri * ri       # NaN positions already have ri = 0

      # ---- Weighted least-squares linear fit (charLowess.m lines 133-147) --
      sw   <- sum(w)
      if (sw < .Machine$double.eps) { ys[i] <- NA_real_; next }

      # Replace NaN yi with 0 for arithmetic (weight is already 0 so they
      # contribute nothing to the sums, but NA would propagate otherwise).
      yi_safe <- yi;  yi_safe[nan_in_win] <- 0.0

      swx  <- sum(w * xi)
      swy  <- sum(w * yi_safe)
      swx2 <- sum(w * xi^2)
      swxy <- sum(w * xi * yi_safe)
      det  <- sw * swx2 - swx^2

      if (abs(det) < .Machine$double.eps * max(abs(c(sw * swx2, swx^2)))) {
        ys[i] <- swy / sw                            # near-singular: mean
      } else {
        a     <- (swy * swx2 - swx * swxy) / det
        b     <- (sw  * swxy - swx * swy)  / det
        ys[i] <- a + b * x[i]
      }
    }

    # ---- Bisquare robustness weights (charLowess.m lines 151-157) ----------
    # Compute the scale (median absolute residual) only from valid positions,
    # matching MATLAB smooth() which skips NaN when computing the median.
    if (iter > 0L && it < n_iter) {
      res    <- y - ys
      s_val  <- stats::median(abs(res[!nan_mask]), na.rm = TRUE)
      if (is.na(s_val) || s_val < .Machine$double.eps) break
      u      <- res / (6.0 * s_val)
      new_rw <- pmax(0.0, 1.0 - u^2)^2 * (abs(u) < 1.0)
      new_rw[nan_mask] <- 0.0       # keep NaN positions excluded
      rw <- new_rw
    }
  }

  ys
}
